service/src/event/service.rs

```diff
@@ -1,27 +1,31 @@
-use std::collections::{HashMap, HashSet};
-
 use async_trait::async_trait;
 use ceramic_core::{EventId, Network};
 use ceramic_event::unvalidated;
 use ceramic_store::{CeramicOneEvent, EventInsertable, EventInsertableBody, SqlitePool};
 use cid::Cid;
 use futures::Stream;
 use ipld_core::ipld::Ipld;
-use recon::ReconItem;
 use tracing::{trace, warn};
 
 use super::{
     migration::Migrator,
-    ordering_task::{
-        DeliverableEvent, DeliverableMetadata, DeliverableTask, DeliveredEvent, OrderingState,
-        OrderingTask, StreamEvents,
-    },
+    order_events::OrderEvents,
+    ordering_task::{DeliverableTask, OrderingTask},
 };
 
 use crate::{Error, Result};
 
+/// How many events to select at once to see if they've become deliverable when we have downtime
+/// Used at startup and occassionally in case we ever dropped something
+/// We keep the number small for now as we may need to traverse many prevs for each one of these and load them into memory.
+const DELIVERABLE_EVENTS_BATCH_SIZE: u32 = 1000;
+
+/// How many batches of undelivered events are we willing to process on start up?
+/// To avoid an infinite loop. It's going to take a long time to process `DELIVERABLE_EVENTS_BATCH_SIZE * MAX_ITERATIONS` events
```

Comment:
this means we can process up to 1 billion undelivered events at startup.

Should we add an upfront check to fail fast if the count of undelivered events is more than this when the node starts up?

Reply:
It'd might be good to log a warning, I was considering whether adding some time estimation log based on how many undelivered events exist would be useful but decided against it. Sqlite does a full table scan to get a count result, so hesitated to add the overhead rather than just make an attempt and inform the user with some messages as we go.

I don't think it's worth failing, because we could process 1 billion and they could restart and process 1 billion more... and if we actually get to this situation things are in a really bad state since we'd need to have like a TB of unordered events and we're probably being spammed somehow.

Anyway, I'm open to making a change or adding a ticket to come back to it, generally, I think making a ticket about startup processing (migrations, ordering, build redtree, etc) might be worthwhile and we can make sure we're offloading what is reasonable, blocking what we must, and being informative about what we're doing.

Reply:
created AES-177